File: ex2.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def swap_pairs_main_recursion(head: ListNode) -> ListNode:
    if not head or not head.next:
        return head

    next_node = head.next

    head.next = swap_pairs_main_recursion(next_node.next)
    next_node.next = head
    return next_node


def swap_pairs_tail_recursion(head: ListNode, prev: ListNode = None) -> ListNode:
    if not head or not head.next:
        return head

    next_node = head.next

    head.next = swap_pairs_tail_recursion(next_node.next, head)
    next_node.next = head

    if prev is None:
        return next_node

    prev.next = next_node

    return head


head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
new_head_main = swap_pairs_main_recursion(head)
new_head_tail = swap_pairs_tail_recursion(head)

File: recursion/ex2.cs
public class ListNode
{
    public int val;
    public ListNode next;

    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}

public class Solution
{
    public ListNode SwapPairs(ListNode head)
    {
        if (head == null || head.next == null)
            return head;

        ListNode firstNode = head;
        ListNode secondNode = head.next;

        firstNode.next = SwapPairs(secondNode.next);
        secondNode.next = firstNode;

        return secondNode;
    }
}
